go/ephemeral/user_ek.go

```diff
@@ -52,29 +55,61 @@ func postNewUserEK(ctx context.Context, g *libkb.GlobalContext, sig string, boxe
 	return err
 }
 
-func PublishNewUserEK(ctx context.Context, g *libkb.GlobalContext) (data keybase1.UserEkMetadata, err error) {
+func PublishNewUserEK(ctx context.Context, g *libkb.GlobalContext) (metadata keybase1.UserEkMetadata, err error) {
 	defer g.CTrace(ctx, "PublishNewUserEK", func() error { return err })()
 
 	currentMerkleRoot, err := g.GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
 	if err != nil {
-		return data, err
+		return metadata, err
 	}
 
-	// TODO: Read the actual generation from the userEK store.
-	var generation keybase1.EkGeneration = 1
-
 	seed, err := newUserEphemeralSeed()
 	if err != nil {
-		return data, err
+		return metadata, err
 	}
 
-	// TODO: Store the key.
+	storage := g.GetUserEKBoxStorage()
+	generation, err := storage.MaxGeneration(ctx)
+	if err != nil {
+		// Let's try to get the max from the server
+		g.Log.CDebugf(ctx, "Error getting maxGeneration from storage")
+		activeMetadata, err := GetLatestUserEKMetadataMaybeStale(ctx, g)
+		if err != nil {
+			return metadata, err
+		}
+		generation = activeMetadata.Generation
+	}
+	generation++
+
+	metadata, myUserEKBoxed, err := signAndPublishUserEK(ctx, g, generation, seed, currentMerkleRoot)
+	if err != nil {
+		g.Log.CDebugf(ctx, "Error posting deviceEK, retrying with server maxGeneration")
+		// Let's retry posting with the server given max
+		activeMetadata, err := GetLatestUserEKMetadataMaybeStale(ctx, g)
+		if err != nil {
+			return metadata, err
+		}
+		generation = activeMetadata.Generation
+		generation++
+		metadata, myUserEKBoxed, err = signAndPublishUserEK(ctx, g, generation, seed, currentMerkleRoot)
+	}
+
+	if err != nil {
+		return metadata, err
+	}
```

Comment:
This check only applies to line 94 above, right? It might be clearer to move it up adjacent to that line.

Reply:
yeah agreed